### DISCUSSION TITLE
Should payments work with multi-page forms?

### DISCUSSION THREAD

**Discussion by ben-at-sahmri** [solspace/craft-freeform#225] [Questions & Answers] [answered]
Hi,

I'm working with Freeform + Payments to produce a multi-page donation form with credit card fields as its last page. Including the credit card fields on anything other than the first page of a multi-page form causes the Stripe fields not to be inserted. The form shows the placeholders and labels, but it looks like the Stripe script is not initialized. 

----

_Edit: Adding more info after further digging on a minimal test case_: This appears to be a side-effect of using the `fieldIdPrefix` option for the form, [as specified in the docs](https://docs.solspace.com/craft/freeform/v3/integrations/payments/#requirements) for situations with more than one form on the page (which my production forms will have). 

Within `StripeHandler` the method `_isStripePresent` looks for an element with the id `this.paymentFieldId + '_card_number'` - taking into account the fieldIdPrefix as expected. However, the Stripe form elements don't include the fieldIdPrefix, so the check fails.

<img width="970" alt="fieldIdPrefix-not-included-in-stripe-field-ids" src="https://user-images.githubusercontent.com/70672160/139198708-f27e8465-018f-4615-98b7-a4b7f6ed41ee.png">

Removing the fieldIdPrefix option fixes this check, but presumably the form then won't work if there is more than one Stripe form on the page.

----

_More investigation:_ After looking at the way the field ID is created via `customAttributes`, I can see that on a form with a `fieldIdPrefix`, the prefix is set correctly on the first page when accessed via `getIdAttribute`. After going to the next page via AJAX, calls to `getIdAttribute` yield a `fieldIdPrefix` of null. (Screenshots from my test form, `var_dump`'ing various values):

![abtractfield-php](https://user-images.githubusercontent.com/70672160/139206951-dacc06a6-62a9-4404-bff7-6aa3e7f22618.png)

----

_Still more_: OK, tracing this back, I can see that when the response is sent back via AJAX, the form is found via its ID and rendered independently. In packages/plugin/src/Controllers/ApiController.php#actionForm we see 

Line 57: `$formModel = $this->getFormsService()->getFormById($formId);`

Then in #toAjaxResponse:

Line 548: `'html' => $form->render()`

`fieldIdPrefix` is set in the calling page template, which is a completely separate context to looking up a form by ID and calling `render` on it. This suggests to me that fieldIdPrefix and AJAX aren't (currently) compatible with each other, which in turn means that having multiple payment forms (which require AJAX) on a page isn't supported. Is that correct?

I'm happy to move this to an issue if that's appropriate - I just want to make sure I understand the whole picture.

This is a fresh install using Freeform Pro 3.11.10 in trial mode, and Stripe in test mode.

Thanks!

**Comment by kjmartens** [comment, marked as answer]
Hi @ben-at-sahmri,

Sorry about the trouble here. We've made a large number of improvements to form processing and handling, and I can confirm that this issue exists in Freeform 3.11, but is resolved in Freeform 3.12. v3.12 is currently in beta but will likely be released within the next week or so. Due to the complex nature of the improvements we made (specifically to address an array of issues like this), we cannot carry over these changes to 3.11, etc. If you'd like to try out the beta, you can follow the guide here:

https://docs.solspace.com/craft/freeform/v3/setup/betas.html

**Reply by ben-at-sahmri** [reply]
Thanks @kjmartens. I did figure out a not-too-horrible workaround in the meantime. We're running up to a production launch now, so I won't attempt the beta yet, but will have a look once we get past our launch date.
